src/ripple/app/tx/impl/AMMVote.cpp

```diff
@@ -0,0 +1,239 @@
+        return res;
+    }
+
+    if (ctx.tx.getFlags() & tfUniversalMask)
+    {
+        JLOG(ctx.j.debug()) << "AMM Vote: invalid flags.";
+        return temINVALID_FLAG;
+    }
+
+    if (ctx.tx[sfTradingFee] > TRADING_FEE_THRESHOLD)
+    {
+        JLOG(ctx.j.debug()) << "AMM Vote: invalid trading fee.";
+        return temBAD_FEE;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+AMMVote::preclaim(PreclaimContext const& ctx)
+{
+    if (auto const ammSle =
+            ctx.view.read(keylet::amm(ctx.tx[sfAsset], ctx.tx[sfAsset2]));
+        !ammSle)
+    {
+        JLOG(ctx.j.debug()) << "AMM Vote: Invalid asset pair.";
+        return terNO_AMM;
+    }
+
+    return tesSUCCESS;
+}
+
+static std::pair<TER, bool>
+applyVote(
+    ApplyContext& ctx_,
+    Sandbox& sb,
+    AccountID const& account_,
+    beast::Journal j_)
+{
+    auto const feeNew = ctx_.tx[sfTradingFee];
+    auto ammSle = sb.peek(keylet::amm(ctx_.tx[sfAsset], ctx_.tx[sfAsset2]));
+    if (!ammSle)
+        return {tecINTERNAL, false};
+    STAmount const lptAMMBalance = (*ammSle)[sfLPTokenBalance];
+    auto const lpTokensNew = ammLPHolds(sb, *ammSle, account_, ctx_.journal);
+    if (lpTokensNew == beast::zero)
+    {
+        JLOG(ctx_.journal.debug()) << "AMM Vote: account is not LP.";
+        return {tecAMM_INVALID_TOKENS, false};
+    }
+
+    std::optional<STAmount> minTokens;
+    std::size_t minPos{0};
+    AccountID minAccount{0};
+    std::uint32_t minFee{0};
+    STArray updatedVoteSlots;
+    Number num{0};
+    Number den{0};
+    // Account already has vote entry
+    bool foundAccount = false;
+    // Iterate over the current vote entries and update each entry
+    // per current total tokens balance and each LP tokens balance.
+    // Find the entry with the least tokens and whether the account
+    // has the vote entry.
+    for (auto const& entry : ammSle->getFieldArray(sfVoteSlots))
+    {
+        auto const account = entry[sfAccount];
+        auto lpTokens = ammLPHolds(sb, *ammSle, account, ctx_.journal);
+        if (lpTokens == beast::zero)
+        {
+            JLOG(j_.debug())
+                << "AMMVote::applyVote, account " << account << " is not LP";
+            continue;
+        }
+        auto feeVal = entry[sfTradingFee];
+        STObject newEntry{sfVoteEntry};
+        // The account already has the vote entry.
+        if (account == account_)
+        {
+            lpTokens = lpTokensNew;
+            feeVal = feeNew;
+            foundAccount = true;
+        }
+        // Keep running numerator/denominator to calculate the updated fee.
+        num += feeVal * lpTokens;
+        den += lpTokens;
+        newEntry.setAccountID(sfAccount, account);
+        if (feeVal != 0)
+            newEntry.setFieldU16(sfTradingFee, feeVal);
+        newEntry.setFieldU32(
+            sfVoteWeight,
+            static_cast<std::int64_t>(
+                Number(lpTokens) * 100000 / lptAMMBalance));
```

Comment:
So just to clarify, `sfVoteWeight` is write-only in the transactors, and is only read for the `amm_info` RPC command? In that case, what's the point of storing it? Just so callers can verify the math?